ral-facilities/datagateway-api
# fix: #444 Generate publication dates between two specific dates

This PR will close #444 

## Description
This PR changes how dates are generated for `DataPublication.publicationDate` and `DataPublicationDate.date`. They are now set to a random date between 2008 and 2023 - the previous range was 15 years so this is the same. `DataPublicationDate.date` used to be a different date to the associated `DataPublication.publicationDate` but they are now set as the same date. I did this because of how I read the issue (and my lack of knowledge in how these fields are actually used) but I'm not sure if they should be or not? Let me know if it should be changed back, it's only a one line change.

## Testing Instructions
Check that data generated by this branch works for the e2e tests for DataGateway. I wasn't sure if you were having difficulty with existing e2e tests or new ones so I haven't done this myself yet, but happy to create a branch on DG pointing to this branch and run the CI if that helps (just let me know).

I haven't tested whether the same dates are generated when the data is generated on different days. Here's some output from the database, I'll re-generate the data in two days time and see if it's still the same:

```bash
MariaDB [icatdb]> SELECT ID, PUBLICATIONDATE FROM DATAPUBLICATION LIMIT 15;
+----+---------------------+
| ID | PUBLICATIONDATE     |
+----+---------------------+
|  1 | 2021-03-23 00:00:00 |
|  2 | 2017-10-02 00:00:00 |
|  3 | 2009-02-02 00:00:00 |
|  4 | 2022-02-06 00:00:00 |
|  5 | 2022-12-14 00:00:00 |
|  6 | 2014-11-02 00:00:00 |
|  7 | 2010-03-18 00:00:00 |
|  8 | 2010-02-24 00:00:00 |
|  9 | 2018-12-22 00:00:00 |
| 10 | 2017-03-12 00:00:00 |
| 11 | 2010-12-24 00:00:00 |
| 12 | 2019-06-14 00:00:00 |
| 13 | 2011-05-06 00:00:00 |
| 14 | 2018-02-16 00:00:00 |
| 15 | 2021-12-25 00:00:00 |
+----+---------------------+
15 rows in set (0.00 sec)

MariaDB [icatdb]> SELECT ID, DATE_ FROM DATAPUBLICATIONDATE LIMIT 15;
+----+------------+
| ID | DATE_      |
+----+------------+
|  1 | 2021-03-23 |
|  2 | 2017-10-02 |
|  3 | 2009-02-02 |
|  4 | 2022-02-06 |
|  5 | 2022-12-14 |
|  6 | 2014-11-02 |
|  7 | 2010-03-18 |
|  8 | 2010-02-24 |
|  9 | 2018-12-22 |
| 10 | 2017-03-12 |
| 11 | 2010-12-24 |
| 12 | 2019-06-14 |
| 13 | 2011-05-06 |
| 14 | 2018-02-16 |
| 15 | 2021-12-25 |
+----+------------+
15 rows in set (0.00 sec)
```

Note that one is a datetime and the other is just a date (well, string in ISO 8601 format) - this is how the data types were so I've kept them the same. If they need to be changed, let me know and I can do that.


## Agile Board Tracking
Connect to #444
